tests/integration/tls_tests/test_tls.py

```diff
@@ -36,6 +46,69 @@ async def test_enable_tls(ops_test: OpsTest) -> None:
         assert await check_tls(ops_test, unit, enabled=True)
 
 
+async def test_set_tls_key(ops_test: OpsTest) -> None:
+    """Verify rotating tls private keys restarts mongod with new certificates."""
+    # dict of values for cert file certion and mongod service start times. After resetting the
+    # private keys these certificates should be updated and the mongod service should be
+    # restarted
+    original_tls_times = {}
+    for unit in ops_test.model.applications[DATABASE_APP_NAME].units:
+        original_tls_times[unit.name] = {}
+        original_tls_times[unit.name]["external_cert"] = await time_file_created(
+            ops_test, unit.name, EXTERNAL_CERT_PATH
+        )
+        original_tls_times[unit.name]["internal_cert"] = await time_file_created(
+            ops_test, unit.name, INTERNAL_CERT_PATH
+        )
+        original_tls_times[unit.name]["mongod_service"] = await time_process_started(
+            ops_test, unit.name, DB_SERVICE
+        )
+
+    with open(PRIVATE_KEY_PATH) as f:
+        key_contents = f.readlines()
+        key_contents = "".join(key_contents)
+
+    # set internal key
+    leader_unit = await find_unit(ops_test, leader=True)
+    leader_unit = await find_unit(ops_test, leader=True)
+    await leader_unit.run_action(
+        action_name="set-tls-private-key", params="=".join(["external-key", key_contents])
+    )
+
+    # set external key
+    leader_unit = await find_unit(ops_test, leader=True)
+    await leader_unit.run_action(
+        action_name="set-tls-private-key", params="=".join(["internal-key", key_contents])
+    )
+
+    # sleep for long enough to process new certificate request and update certs
+    sleep(30)
```

Comment:
Is there something you can wait for here? Maybe action completion, or a change in status using `wait_for_idle`?

Reply:
Great question @WRFitch :) Unfortunately I cannot wait for the action to complete since setting the keys doesn't immediately result in the charm enabling TLS. Changing the keys generates a certificate signing requests (CSR), which isn't processed until the CA processes this request. Then the TLS charm will issue a certificate-available-event which after that event is issued we receive the event on the DB charm and restart TLS with the correct certs. (For this reason I cannot wait for idle either 😢 )

I could use retry over a period of 30secs but this would be 5-7 more lines of code and essentially do the same thing so I will keep this as sleep if that is okay.

Reply:
Great, thanks for the explanation :)